**Predicting flight delays in advance with machine learning models**

Flight delays or cancellations have become unavoidable in flight travels. The effects of such delays or cancellations lead to travelers losing some money due to missing cruises, prepaid hotel rooms, increased parking fees, cancellation or postponement of business trips and sometimes the additional cost of unplanned air tickets.Travelers also miss vacation and valuable time with family and friends as a result of these delays.

Building a model that predicts flight delays will serve as an important tool that could be used by all air travelers to predict flight cancellations in advance and to prevent  loss  associated with flight delays or cancellations. It will also give air travelers ample time to replan meetups, reschedule trips, make safer accommodations choices and prevent them from being stranded due to unplanned flight delays.

The data sources  for this project include:
a.  The 2015 Flight Delays and Cancellations (kaggle.com) downloaded from Kaggle. It has about three files airlines.csv, airports.csv and  flights.csv that has information about flight delays in 2015. However, only the flights.csv will be used for this project. This  file has  1048576 rows and 31 columns.
b. 

The second file that will be used for this project is the airlinedelaycauses_DelayedFlights.csv also downloaded from Kaggle. It is made up of similar features as the first file. This data has information about flight delays and cancellations in 2008. It has 1936758 rows and 30 columnsc. .

The supplementary dataset will  be downloaded from Open data @ OurAirports. It is named the airport.csv and has 80559 rows and 18 columns.mns.


The objective of this project will be met by 
1. Cleaning and transforming datasets from data sources using Python tools like NumPy and Pandas.
2. Cleaned datasets will be merged into one Pandas data frame. This data frame will be analyzed using the same tools. Exploratory data analysis will be conducted and detected trends, patterns and relationships between features will be visualized using matplotlib and seaborn.
3. Data features will be preprocessed into forms that can be easily utilized by machine learning algorithms. 
4. Three machine learning models will be built using different types of machine learning algorithms.
5. These models will be evaluated and the best performing model will be chosen as the working model of this project. Models will be improved using cross validations or hyperparameter tuning to get the model in the best performing state



In [1]:
# Loading libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Loading file from 2008

Delay_2008 =pd.read_csv('airlinedelaycauses_DelayedFlights.csv')
Delay_2008.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  

[5 rows x 30 columns]

In [3]:
# Checking the shape of the dataset

Delay_2008.shape

(1936758, 30)

In [4]:
# Checking summary statistics of dataset

Delay_2008.describe()

Unnamed: 0       Year         Month    DayofMonth     DayOfWeek  \
count  1.936758e+06  1936758.0  1.936758e+06  1.936758e+06  1.936758e+06   
mean   3.341651e+06     2008.0  6.111106e+00  1.575347e+01  3.984827e+00   
std    2.066065e+06        0.0  3.482546e+00  8.776272e+00  1.995966e+00   
min    0.000000e+00     2008.0  1.000000e+00  1.000000e+00  1.000000e+00   
25%    1.517452e+06     2008.0  3.000000e+00  8.000000e+00  2.000000e+00   
50%    3.242558e+06     2008.0  6.000000e+00  1.600000e+01  4.000000e+00   
75%    4.972467e+06     2008.0  9.000000e+00  2.300000e+01  6.000000e+00   
max    7.009727e+06     2008.0  1.200000e+01  3.100000e+01  7.000000e+00   

            DepTime    CRSDepTime       ArrTime    CRSArrTime     FlightNum  \
count  1.936758e+06  1.936758e+06  1.929648e+06  1.936758e+06  1.936758e+06   
mean   1.518534e+03  1.467473e+03  1.610141e+03  1.634225e+03  2.184263e+03   
std    4.504853e+02  4.247668e+02  5.481781e+02  4.646347e+02  1.944702e+03   
min    1.000000e+00  0.000000e+00  1.000000e+00  0.000000e+00  1.000000e+00   
25%    1.203000e+03  1.135000e+03  1.316000e+03  1.325000e+03  6.100000e+02   
50%    1.545000e+03  1.510000e+03  1.715000e+03  1.705000e+03  1.543000e+03   
75%    1.900000e+03  1.815000e+03  2.030000e+03  2.014000e+03  3.422000e+03   
max    2.400000e+03  2.359000e+03  2.400000e+03  2.400000e+03  9.742000e+03   

       ...      Distance        TaxiIn       TaxiOut     Cancelled  \
count  ...  1.936758e+06  1.929648e+06  1.936303e+06  1.936758e+06   
mean   ...  7.656862e+02  6.812975e+00  1.823220e+01  3.268348e-04   
std    ...  5.744797e+02  5.273595e+00  1.433853e+01  1.807562e-02   
min    ...  1.100000e+01  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  3.380000e+02  4.000000e+00  1.000000e+01  0.000000e+00   
50%    ...  6.060000e+02  6.000000e+00  1.400000e+01  0.000000e+00   
75%    ...  9.980000e+02  8.000000e+00  2.100000e+01  0.000000e+00   
max    ...  4.962000e+03  2.400000e+02  4.220000e+02  1.000000e+00   

           Diverted  CarrierDelay  WeatherDelay      NASDelay  SecurityDelay  \
count  1.936758e+06  1.247488e+06  1.247488e+06  1.247488e+06   1.247488e+06   
mean   4.003598e-03  1.917940e+01  3.703571e+00  1.502164e+01   9.013714e-02   
std    6.314722e-02  4.354621e+01  2.149290e+01  3.383305e+01   2.022714e+00   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
50%    0.000000e+00  2.000000e+00  0.000000e+00  2.000000e+00   0.000000e+00   
75%    0.000000e+00  2.100000e+01  0.000000e+00  1.500000e+01   0.000000e+00   
max    1.000000e+00  2.436000e+03  1.352000e+03  1.357000e+03   3.920000e+02   

       LateAircraftDelay  
count       1.247488e+06  
mean        2.529647e+01  
std         4.205486e+01  
min         0.000000e+00  
25%         0.000000e+00  
50%         8.000000e+00  
75%         3.300000e+01  
max         1.316000e+03  

[8 rows x 25 columns]

In [5]:
# Checking the data types of columns

Delay_2008.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

In [6]:
# Checking missing values

Delay_2008.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

In [7]:
# Replacing missing values with 0

Delay_2008 = Delay_2008.fillna('0.0')

In [8]:
# Dropping irrevalant columns

Delay_2008 = Delay_2008.drop(columns=['CancellationCode'],axis =1)

In [9]:
# Ensuring there are no missing values

Delay_2008.isnull().sum()

Unnamed: 0           0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

In [10]:
Delay_2008.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

  ArrTime  CRSArrTime UniqueCarrier  ...  Distance TaxiIn TaxiOut Cancelled  \
0  2211.0        2225            WN  ...       810    4.0     8.0         0   
1  1002.0        1000            WN  ...       810    5.0    10.0         0   
2   804.0         750            WN  ...       515    3.0    17.0         0   
3  1959.0        1925            WN  ...       515    3.0    10.0         0   
4  2121.0        2110            WN  ...       688    4.0    10.0         0   

  Diverted CarrierDelay  WeatherDelay NASDelay SecurityDelay  \
0        0          0.0           0.0      0.0           0.0   
1        0          0.0           0.0      0.0           0.0   
2        0          0.0           0.0      0.0           0.0   
3        0          2.0           0.0      0.0           0.0   
4        0          0.0           0.0      0.0           0.0   

   LateAircraftDelay  
0                0.0  
1                0.0  
2                0.0  
3               32.0  
4                0.0  

[5 rows x 29 columns]

Cleaning dataset from 2015

In [11]:
# Loading file from 2015

Delay_2015 = pd.read_csv('flights.csv')
Delay_2015.head()

C:\Users\sonia\AppData\Local\Temp\ipykernel_2364\1282143124.py:3: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  Delay_2015 = pd.read_csv('flights.csv')


YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AA           2336      N3KUAA   
2  2015      1    1            4      US            840      N171US   
3  2015      1    1            4      AA            258      N3HYAA   
4  2015      1    1            4      AS            135      N527AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  ARRIVAL_TIME  \
0            ANC                 SEA                    5  ...         408.0   
1            LAX                 PBI                   10  ...         741.0   
2            SFO                 CLT                   20  ...         811.0   
3            LAX                 MIA                   20  ...         756.0   
4            SEA                 ANC                   25  ...         259.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          -22.0         0          0                  NaN               NaN   
1           -9.0         0          0                  NaN               NaN   
2            5.0         0          0                  NaN               NaN   
3           -9.0         0          0                  NaN               NaN   
4          -21.0         0          0                  NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0             NaN            NaN                  NaN            NaN  
1             NaN            NaN                  NaN            NaN  
2             NaN            NaN                  NaN            NaN  
3             NaN            NaN                  NaN            NaN  
4             NaN            NaN                  NaN            NaN  

[5 rows x 31 columns]

Some of the column seem similar to others so further ivestigation was required to find out what columns represent.

It turns out the departure time represents the time the airline takes leaves the gate and the wheel off time is the time the airline takes off from runway of the airport. The arrival time is the time the airline arrives at the gate and the wheels on time is the time the airline arrives or touches ground of the airport

In [12]:
Delay_2015[['DEPARTURE_TIME', 'WHEELS_OFF','ARRIVAL_TIME','WHEELS_ON']]

DEPARTURE_TIME  WHEELS_OFF  ARRIVAL_TIME  WHEELS_ON
0                2354.0        15.0         408.0      404.0
1                   2.0        14.0         741.0      737.0
2                  18.0        34.0         811.0      800.0
3                  15.0        30.0         756.0      748.0
4                  24.0        35.0         259.0      254.0
...                 ...         ...           ...        ...
5819074          2355.0        17.0         753.0      749.0
5819075          2355.0        12.0         430.0      427.0
5819076          2350.0         7.0         432.0      424.0
5819077          2353.0         3.0         330.0      327.0
5819078            14.0        28.0         442.0      437.0

[5819079 rows x 4 columns]

In [13]:
# Checking the shape of dataset

Delay_2015.shape

(5819079, 31)

In [14]:
# Checking summary statistics

Delay_2015.describe()

YEAR         MONTH           DAY   DAY_OF_WEEK  FLIGHT_NUMBER  \
count  5819079.0  5.819079e+06  5.819079e+06  5.819079e+06   5.819079e+06   
mean      2015.0  6.524085e+00  1.570459e+01  3.926941e+00   2.173093e+03   
std          0.0  3.405137e+00  8.783425e+00  1.988845e+00   1.757064e+03   
min       2015.0  1.000000e+00  1.000000e+00  1.000000e+00   1.000000e+00   
25%       2015.0  4.000000e+00  8.000000e+00  2.000000e+00   7.300000e+02   
50%       2015.0  7.000000e+00  1.600000e+01  4.000000e+00   1.690000e+03   
75%       2015.0  9.000000e+00  2.300000e+01  6.000000e+00   3.230000e+03   
max       2015.0  1.200000e+01  3.100000e+01  7.000000e+00   9.855000e+03   

       SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY      TAXI_OUT  \
count         5.819079e+06    5.732926e+06     5.732926e+06  5.730032e+06   
mean          1.329602e+03    1.335204e+03     9.370158e+00  1.607166e+01   
std           4.837518e+02    4.964233e+02     3.708094e+01  8.895574e+00   
min           1.000000e+00    1.000000e+00    -8.200000e+01  1.000000e+00   
25%           9.170000e+02    9.210000e+02    -5.000000e+00  1.100000e+01   
50%           1.325000e+03    1.330000e+03    -2.000000e+00  1.400000e+01   
75%           1.730000e+03    1.740000e+03     7.000000e+00  1.900000e+01   
max           2.359000e+03    2.400000e+03     1.988000e+03  2.250000e+02   

         WHEELS_OFF  ...  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
count  5.730032e+06  ...       5.819079e+06  5.726566e+06   5.714008e+06   
mean   1.357171e+03  ...       1.493808e+03  1.476491e+03   4.407057e+00   
std    4.980094e+02  ...       5.071647e+02  5.263197e+02   3.927130e+01   
min    1.000000e+00  ...       1.000000e+00  1.000000e+00  -8.700000e+01   
25%    9.350000e+02  ...       1.110000e+03  1.059000e+03  -1.300000e+01   
50%    1.343000e+03  ...       1.520000e+03  1.512000e+03  -5.000000e+00   
75%    1.754000e+03  ...       1.918000e+03  1.917000e+03   8.000000e+00   
max    2.400000e+03  ...       2.400000e+03  2.400000e+03   1.971000e+03   

           DIVERTED     CANCELLED  AIR_SYSTEM_DELAY  SECURITY_DELAY  \
count  5.819079e+06  5.819079e+06      1.063439e+06    1.063439e+06   
mean   2.609863e-03  1.544643e-02      1.348057e+01    7.615387e-02   
std    5.102012e-02  1.233201e-01      2.800368e+01    2.143460e+00   
min    0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
25%    0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
50%    0.000000e+00  0.000000e+00      2.000000e+00    0.000000e+00   
75%    0.000000e+00  0.000000e+00      1.800000e+01    0.000000e+00   
max    1.000000e+00  1.000000e+00      1.134000e+03    5.730000e+02   

       AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
count   1.063439e+06         1.063439e+06   1.063439e+06  
mean    1.896955e+01         2.347284e+01   2.915290e+00  
std     4.816164e+01         4.319702e+01   2.043334e+01  
min     0.000000e+00         0.000000e+00   0.000000e+00  
25%     0.000000e+00         0.000000e+00   0.000000e+00  
50%     2.000000e+00         3.000000e+00   0.000000e+00  
75%     1.900000e+01         2.900000e+01   0.000000e+00  
max     1.971000e+03         1.331000e+03   1.211000e+03  

[8 rows x 26 columns]

In [15]:
# Checking data tyoes of columns

Delay_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5819079 entries, 0 to 5819078
Data columns (total 31 columns):
 #   Column               Dtype  
---  ------               -----  
 0   YEAR                 int64  
 1   MONTH                int64  
 2   DAY                  int64  
 3   DAY_OF_WEEK          int64  
 4   AIRLINE              object 
 5   FLIGHT_NUMBER        int64  
 6   TAIL_NUMBER          object 
 7   ORIGIN_AIRPORT       object 
 8   DESTINATION_AIRPORT  object 
 9   SCHEDULED_DEPARTURE  int64  
 10  DEPARTURE_TIME       float64
 11  DEPARTURE_DELAY      float64
 12  TAXI_OUT             float64
 13  WHEELS_OFF           float64
 14  SCHEDULED_TIME       float64
 15  ELAPSED_TIME         float64
 16  AIR_TIME             float64
 17  DISTANCE             int64  
 18  WHEELS_ON            float64
 19  TAXI_IN              float64
 20  SCHEDULED_ARRIVAL    int64  
 21  ARRIVAL_TIME         float64
 22  ARRIVAL_DELAY        float64
 23  DIVERTED             int64  
 24

In [16]:
# Checking missing values

Delay_2015.isnull().sum()

YEAR                         0
MONTH                        0
DAY                          0
DAY_OF_WEEK                  0
AIRLINE                      0
FLIGHT_NUMBER                0
TAIL_NUMBER              14721
ORIGIN_AIRPORT               0
DESTINATION_AIRPORT          0
SCHEDULED_DEPARTURE          0
DEPARTURE_TIME           86153
DEPARTURE_DELAY          86153
TAXI_OUT                 89047
WHEELS_OFF               89047
SCHEDULED_TIME               6
ELAPSED_TIME            105071
AIR_TIME                105071
DISTANCE                     0
WHEELS_ON                92513
TAXI_IN                  92513
SCHEDULED_ARRIVAL            0
ARRIVAL_TIME             92513
ARRIVAL_DELAY           105071
DIVERTED                     0
CANCELLED                    0
CANCELLATION_REASON    5729195
AIR_SYSTEM_DELAY       4755640
SECURITY_DELAY         4755640
AIRLINE_DELAY          4755640
LATE_AIRCRAFT_DELAY    4755640
WEATHER_DELAY          4755640
dtype: int64

In [17]:
# Replacing missing values with 0

Delay_2015 = Delay_2015.fillna('0.0')

In [18]:
# Ensuring there are no missing values

Delay_2015.isnull().sum()

YEAR                   0
MONTH                  0
DAY                    0
DAY_OF_WEEK            0
AIRLINE                0
FLIGHT_NUMBER          0
TAIL_NUMBER            0
ORIGIN_AIRPORT         0
DESTINATION_AIRPORT    0
SCHEDULED_DEPARTURE    0
DEPARTURE_TIME         0
DEPARTURE_DELAY        0
TAXI_OUT               0
WHEELS_OFF             0
SCHEDULED_TIME         0
ELAPSED_TIME           0
AIR_TIME               0
DISTANCE               0
WHEELS_ON              0
TAXI_IN                0
SCHEDULED_ARRIVAL      0
ARRIVAL_TIME           0
ARRIVAL_DELAY          0
DIVERTED               0
CANCELLED              0
CANCELLATION_REASON    0
AIR_SYSTEM_DELAY       0
SECURITY_DELAY         0
AIRLINE_DELAY          0
LATE_AIRCRAFT_DELAY    0
WEATHER_DELAY          0
dtype: int64

In [19]:
Delay_2015.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AA           2336      N3KUAA   
2  2015      1    1            4      US            840      N171US   
3  2015      1    1            4      AA            258      N3HYAA   
4  2015      1    1            4      AS            135      N527AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ... ARRIVAL_TIME  \
0            ANC                 SEA                    5  ...        408.0   
1            LAX                 PBI                   10  ...        741.0   
2            SFO                 CLT                   20  ...        811.0   
3            LAX                 MIA                   20  ...        756.0   
4            SEA                 ANC                   25  ...        259.0   

  ARRIVAL_DELAY DIVERTED CANCELLED CANCELLATION_REASON AIR_SYSTEM_DELAY  \
0         -22.0        0         0                 0.0              0.0   
1          -9.0        0         0                 0.0              0.0   
2           5.0        0         0                 0.0              0.0   
3          -9.0        0         0                 0.0              0.0   
4         -21.0        0         0                 0.0              0.0   

  SECURITY_DELAY  AIRLINE_DELAY LATE_AIRCRAFT_DELAY WEATHER_DELAY  
0            0.0            0.0                 0.0           0.0  
1            0.0            0.0                 0.0           0.0  
2            0.0            0.0                 0.0           0.0  
3            0.0            0.0                 0.0           0.0  
4            0.0            0.0                 0.0           0.0  

[5 rows x 31 columns]

Renaming columns from cleaned 2008 dataset(Delay_2008) to match columns from cleaned 2015 dataset( Delay_2015) prior to joing the two datasets

In [20]:
# Renaming columns in Delay_2008 to match columns in Delay_2015

Delay_2008.rename(columns={'Year':'YEAR','Month':'MONTH','DayOfMonth':'DAY','DayOfWeek':'DAY_OF_WEEK','DepTime':'DEPARTURE_TIME','CRSDepTime':'SCHEDULED_DEPARTURE','ArrTime':'ARRIVAL_TIME','CRSArrTime':'SCHEDULED_ARRIVAL','UniqueCarrier':'AIRLINE','FlightNum':'FLIGHT_NUMBER','TailNum':'TAIL_NUMBER','ActualElapsedTime':'ELAPSED_TIME','CRSElapsedTime':'SCHEDULED_TIME','AirTime':'AIR_TIME','ArrDelay':'ARRIVAL_DELAY','DepDelay':'DEPARTURE_DELAY','Origin':'ORIGIN_AIRPORT','Dest':'DESTINATION_AIRPORT','Distance':'DISTANCE','TaxiIn':'TAXI_IN','TaxiOut':'TAXI_OUT','Cancelled':'CANCELLED','Diverted':'DIVERTED','CarrierDelay':'AIR_SYSTEM_DELAY','WeatherDelay': 'WEATHER_DELAY','NASDelay':'NAS_DELAY','SecurityDelay': 'SECURITY_DELAY','LateAircraftDelay':'LATE_AIRCRAFT_DELAY','COUNTRY':'COUNTRY'},inplace=True)

In [21]:
Delay_2008.head()

Unnamed: 0  YEAR  MONTH  DayofMonth  DAY_OF_WEEK  DEPARTURE_TIME  \
0           0  2008      1           3            4          2003.0   
1           1  2008      1           3            4           754.0   
2           2  2008      1           3            4           628.0   
3           4  2008      1           3            4          1829.0   
4           5  2008      1           3            4          1940.0   

   SCHEDULED_DEPARTURE ARRIVAL_TIME  SCHEDULED_ARRIVAL AIRLINE  ...  DISTANCE  \
0                 1955       2211.0               2225      WN  ...       810   
1                  735       1002.0               1000      WN  ...       810   
2                  620        804.0                750      WN  ...       515   
3                 1755       1959.0               1925      WN  ...       515   
4                 1915       2121.0               2110      WN  ...       688   

  TAXI_IN TAXI_OUT CANCELLED DIVERTED AIR_SYSTEM_DELAY  WEATHER_DELAY  \
0     4.0      8.0         0        0              0.0            0.0   
1     5.0     10.0         0        0              0.0            0.0   
2     3.0     17.0         0        0              0.0            0.0   
3     3.0     10.0         0        0              2.0            0.0   
4     4.0     10.0         0        0              0.0            0.0   

  NAS_DELAY SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0       0.0            0.0                  0.0  
1       0.0            0.0                  0.0  
2       0.0            0.0                  0.0  
3       0.0            0.0                 32.0  
4       0.0            0.0                  0.0  

[5 rows x 29 columns]

Joining datasets 

In [22]:
# Join Delay_2008 on top of Delay_2015 and call it Delays

Delays = pd.concat([Delay_2008, Delay_2015],ignore_index=True)
Delays.head()

Unnamed: 0  YEAR  MONTH  DayofMonth  DAY_OF_WEEK DEPARTURE_TIME  \
0         0.0  2008      1         3.0            4         2003.0   
1         1.0  2008      1         3.0            4          754.0   
2         2.0  2008      1         3.0            4          628.0   
3         4.0  2008      1         3.0            4         1829.0   
4         5.0  2008      1         3.0            4         1940.0   

   SCHEDULED_DEPARTURE ARRIVAL_TIME  SCHEDULED_ARRIVAL AIRLINE  ...  \
0                 1955       2211.0               2225      WN  ...   
1                  735       1002.0               1000      WN  ...   
2                  620        804.0                750      WN  ...   
3                 1755       1959.0               1925      WN  ...   
4                 1915       2121.0               2110      WN  ...   

   AIR_SYSTEM_DELAY WEATHER_DELAY NAS_DELAY SECURITY_DELAY  \
0               0.0           0.0       0.0            0.0   
1               0.0           0.0       0.0            0.0   
2               0.0           0.0       0.0            0.0   
3               2.0           0.0       0.0            0.0   
4               0.0           0.0       0.0            0.0   

  LATE_AIRCRAFT_DELAY DAY WHEELS_OFF WHEELS_ON CANCELLATION_REASON  \
0                 0.0 NaN        NaN       NaN                 NaN   
1                 0.0 NaN        NaN       NaN                 NaN   
2                 0.0 NaN        NaN       NaN                 NaN   
3                32.0 NaN        NaN       NaN                 NaN   
4                 0.0 NaN        NaN       NaN                 NaN   

   AIRLINE_DELAY  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

[5 rows x 34 columns]

In [23]:
# Checking the shape of the new dataset

Delays.shape

(7755837, 34)

In [24]:
# Dropping irrelevant columns

Delays= Delays.drop(columns= ['Unnamed: 0'], axis=1)

In [25]:
# Replacing abbreviations in airline column with full airline names from airline.csv

full_airlines = pd.read_csv('airlines.csv')

In [26]:
abbreviations = full_airlines.set_index('IATA_CODE')['AIRLINE'].to_dict()

In [27]:
Delays['AIRLINE'] = Delays['AIRLINE'].replace(abbreviations)
Delays['AIRLINE']

0          Southwest Airlines Co.
1          Southwest Airlines Co.
2          Southwest Airlines Co.
3          Southwest Airlines Co.
4          Southwest Airlines Co.
                    ...          
7755832           JetBlue Airways
7755833           JetBlue Airways
7755834           JetBlue Airways
7755835           JetBlue Airways
7755836           JetBlue Airways
Name: AIRLINE, Length: 7755837, dtype: object

In [28]:
# Repalcing abbreviations of airports with full airport names from airports.csv

full_airports= pd.read_csv('airports.csv')

In [29]:
abbre= full_airports.set_index('IATA_CODE')['AIRPORT'].to_dict()

In [30]:
Delays['ORIGIN_AIRPORT']= Delays['ORIGIN_AIRPORT'].replace(abbre)
Delays['DESTINATION_AIRPORT']= Delays['DESTINATION_AIRPORT'].replace(abbre)

In [31]:
Delays['ORIGIN_AIRPORT']

0                    Washington Dulles International Airport
1                    Washington Dulles International Airport
2                         Indianapolis International Airport
3                         Indianapolis International Airport
4                         Indianapolis International Airport
                                 ...                        
7755832                    Los Angeles International Airport
7755833    John F. Kennedy International Airport (New Yor...
7755834    John F. Kennedy International Airport (New Yor...
7755835                        Orlando International Airport
7755836    John F. Kennedy International Airport (New Yor...
Name: ORIGIN_AIRPORT, Length: 7755837, dtype: object

In [32]:
Delays['DESTINATION_AIRPORT']

0                               Tampa International Airport
1                               Tampa International Airport
2                Baltimore-Washington International Airport
3                Baltimore-Washington International Airport
4                        Jacksonville International Airport
                                 ...                       
7755832    Gen. Edward Lawrence Logan International Airport
7755833                                   Mercedita Airport
7755834              Luis Muñoz Marín International Airport
7755835              Luis Muñoz Marín International Airport
7755836                            Rafael Hernández Airport
Name: DESTINATION_AIRPORT, Length: 7755837, dtype: object

Loading and cleaning supplementary dataset which contain information about airports.

In [33]:
# loading supplementary file

Airports = pd.read_csv("airports supplementary dataset.csv")
Airports.head()

id ident           type                  name  latitude_deg  \
0    6523   00A       heliport     Total RF Heliport     40.070985   
1  323361  00AA  small_airport  Aero B Ranch Airport     38.704022   
2    6524  00AK  small_airport          Lowell Field     59.947733   
3    6525  00AL  small_airport          Epps Airpark     34.864799   
4  506791  00AN  small_airport  Katmai Lodge Airport     59.093287   

   longitude_deg  elevation_ft continent iso_country iso_region  municipality  \
0     -74.933689          11.0       NaN          US      US-PA      Bensalem   
1    -101.473911        3435.0       NaN          US      US-KS         Leoti   
2    -151.692524         450.0       NaN          US      US-AK  Anchor Point   
3     -86.770302         820.0       NaN          US      US-AL       Harvest   
4    -156.456699          80.0       NaN          US      US-AK   King Salmon   

  scheduled_service gps_code iata_code local_code  \
0                no     K00A       NaN        00A   
1                no     00AA       NaN       00AA   
2                no     00AK       NaN       00AK   
3                no     00AL       NaN       00AL   
4                no     00AN       NaN       00AN   

                                           home_link wikipedia_link keywords  
0  https://www.penndot.pa.gov/TravelInPA/airports...            NaN      NaN  
1                                                NaN            NaN      NaN  
2                                                NaN            NaN      NaN  
3                                                NaN            NaN      NaN  
4                                                NaN            NaN      NaN

In [34]:
Airports.shape

(80557, 18)

In [35]:
Airports.columns

Index(['id', 'ident', 'type', 'name', 'latitude_deg', 'longitude_deg',
       'elevation_ft', 'continent', 'iso_country', 'iso_region',
       'municipality', 'scheduled_service', 'gps_code', 'iata_code',
       'local_code', 'home_link', 'wikipedia_link', 'keywords'],
      dtype='object')

In [36]:
Airports.describe()

id  latitude_deg  longitude_deg  elevation_ft
count   80557.000000  80557.000000   80557.000000  65825.000000
mean   181717.925953     26.020670     -28.537423   1304.678207
std    180251.374647     26.141606      85.442239   1671.978083
min         2.000000    -89.989444    -179.876999  -1266.000000
25%     20291.000000     13.239778     -93.719398    213.000000
50%     43103.000000     35.214500     -67.792099    735.000000
75%    339292.000000     43.029277      22.756878   1616.000000
max    541036.000000     82.750000     179.975700  17372.000000

In [37]:
Airports.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80557 entries, 0 to 80556
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 80557 non-null  int64  
 1   ident              80557 non-null  object 
 2   type               80557 non-null  object 
 3   name               80557 non-null  object 
 4   latitude_deg       80557 non-null  float64
 5   longitude_deg      80557 non-null  float64
 6   elevation_ft       65825 non-null  float64
 7   continent          42244 non-null  object 
 8   iso_country        80297 non-null  object 
 9   iso_region         80557 non-null  object 
 10  municipality       75353 non-null  object 
 11  scheduled_service  80557 non-null  object 
 12  gps_code           42572 non-null  object 
 13  iata_code          9093 non-null   object 
 14  local_code         34098 non-null  object 
 15  home_link          4205 non-null   object 
 16  wikipedia_link     164

In [38]:
Airports.isnull().sum()

id                       0
ident                    0
type                     0
name                     0
latitude_deg             0
longitude_deg            0
elevation_ft         14732
continent            38313
iso_country            260
iso_region               0
municipality          5204
scheduled_service        0
gps_code             37985
iata_code            71464
local_code           46459
home_link            76352
wikipedia_link       64149
keywords             61592
dtype: int64

Dropping columns:
1. gps_code column beacause it has missing values but it also has the same information as the ident column which has no missing values
2. home_link, wikipedia_link because information about where data was scrapped is not necessary
3. local_code because it has missing values but it same information as ident column 
4. iata_code because is has a lot of of values missing which can not be replaced
5. iso_country because it has missing values but missing information is found in iso_region
6. municipality because it is irrelevant. iso_region has the required information about where the airport is located
7. scheduled_service because its is also no relevant to this project.

In [39]:
Airports = Airports.drop(columns=['gps_code', 'home_link','wikipedia_link','keywords','local_code','iata_code','municipality','scheduled_service'],axis=1)

In [40]:
Airports.head()

id ident           type                  name  latitude_deg  \
0    6523   00A       heliport     Total RF Heliport     40.070985   
1  323361  00AA  small_airport  Aero B Ranch Airport     38.704022   
2    6524  00AK  small_airport          Lowell Field     59.947733   
3    6525  00AL  small_airport          Epps Airpark     34.864799   
4  506791  00AN  small_airport  Katmai Lodge Airport     59.093287   

   longitude_deg  elevation_ft continent iso_country iso_region  
0     -74.933689          11.0       NaN          US      US-PA  
1    -101.473911        3435.0       NaN          US      US-KS  
2    -151.692524         450.0       NaN          US      US-AK  
3     -86.770302         820.0       NaN          US      US-AL  
4    -156.456699          80.0       NaN          US      US-AK

In [41]:
Airports['iso_country'].value_counts()

iso_country
US    31510
BR     7173
JP     3555
CA     3126
AU     2675
      ...  
GM        1
GS        1
NU        1
GI        1
NF        1
Name: count, Length: 244, dtype: int64

In [42]:
# Checking the values in the conitinent column reveals that North America(NA) is missing. 
# Confirming the continent from the country column is was observed that North America was categorised as 'NaN'

Airports['continent'].value_counts()

continent
AS    11612
EU    11487
SA    11046
OC     4040
AF     4016
AN       43
Name: count, dtype: int64

In [43]:
# Filling missing values in continent with the North America label(NA)

Airports['continent']= Airports['continent'].fillna('NA')

In [44]:
# Checking whether change has been effected

Airports.head()

id ident           type                  name  latitude_deg  \
0    6523   00A       heliport     Total RF Heliport     40.070985   
1  323361  00AA  small_airport  Aero B Ranch Airport     38.704022   
2    6524  00AK  small_airport          Lowell Field     59.947733   
3    6525  00AL  small_airport          Epps Airpark     34.864799   
4  506791  00AN  small_airport  Katmai Lodge Airport     59.093287   

   longitude_deg  elevation_ft continent iso_country iso_region  
0     -74.933689          11.0        NA          US      US-PA  
1    -101.473911        3435.0        NA          US      US-KS  
2    -151.692524         450.0        NA          US      US-AK  
3     -86.770302         820.0        NA          US      US-AL  
4    -156.456699          80.0        NA          US      US-AK

In [45]:
# Checking missing values in the dataframe

Airports.isnull().sum()

id                   0
ident                0
type                 0
name                 0
latitude_deg         0
longitude_deg        0
elevation_ft     14732
continent            0
iso_country        260
iso_region           0
dtype: int64

In [46]:
# Replacing missing values in iso_country column with country value from iso_region column

for col in Airports.columns:
    if col+'_suffix' in Airports.columns:
        Airports[col] = Airports[col].fillna(Airports[col+'_suffix'])

In [47]:
Airports['iso_country']= Airports['iso_country'].fillna(Airports['iso_region'])

In [48]:
Airports.head(5)

id ident           type                  name  latitude_deg  \
0    6523   00A       heliport     Total RF Heliport     40.070985   
1  323361  00AA  small_airport  Aero B Ranch Airport     38.704022   
2    6524  00AK  small_airport          Lowell Field     59.947733   
3    6525  00AL  small_airport          Epps Airpark     34.864799   
4  506791  00AN  small_airport  Katmai Lodge Airport     59.093287   

   longitude_deg  elevation_ft continent iso_country iso_region  
0     -74.933689          11.0        NA          US      US-PA  
1    -101.473911        3435.0        NA          US      US-KS  
2    -151.692524         450.0        NA          US      US-AK  
3     -86.770302         820.0        NA          US      US-AL  
4    -156.456699          80.0        NA          US      US-AK

In [49]:
# Checking number of missing values in iso_country column

Airports['iso_country'].isnull().sum()

0

In [50]:
# Handling missing values in elevation_ft column by replacing missing values with median value

Airports['elevation_ft']= Airports['elevation_ft'].fillna(Airports['elevation_ft'].median())

In [51]:
# Checking number of missing values in elevation_ft column

Airports['elevation_ft'].isnull().sum()

0

In [52]:
# Ensuring there are no missing values

Airports.isnull().sum()

id               0
ident            0
type             0
name             0
latitude_deg     0
longitude_deg    0
elevation_ft     0
continent        0
iso_country      0
iso_region       0
dtype: int64

Data from Delay_2008 and Delay_2015 are about US flight. So the supplementary data must also reflect similar data from only US airports.

In [53]:
Airports = Airports[Airports['iso_country']== 'US']

In [54]:
# Dropping the continent column because its no longer relevant

Airports= Airports.drop(columns=['continent'], axis=1)

In [55]:
Airports.head()

id ident           type                  name  latitude_deg  \
0    6523   00A       heliport     Total RF Heliport     40.070985   
1  323361  00AA  small_airport  Aero B Ranch Airport     38.704022   
2    6524  00AK  small_airport          Lowell Field     59.947733   
3    6525  00AL  small_airport          Epps Airpark     34.864799   
4  506791  00AN  small_airport  Katmai Lodge Airport     59.093287   

   longitude_deg  elevation_ft iso_country iso_region  
0     -74.933689          11.0          US      US-PA  
1    -101.473911        3435.0          US      US-KS  
2    -151.692524         450.0          US      US-AK  
3     -86.770302         820.0          US      US-AL  
4    -156.456699          80.0          US      US-AK

Now that all datasets are clean, they are will be merged into one dataset, starting with merging Delay_2008 on top of Delay_2015.

In [56]:
Airports.dtypes

id                 int64
ident             object
type              object
name              object
latitude_deg     float64
longitude_deg    float64
elevation_ft     float64
iso_country       object
iso_region        object
dtype: object

After all datsets have  been cleaned. Airport dataset will be merged with the initial Delays dataset and the final dataset will be called fl_delay.

In [57]:
# Merge Delays with supplementary dataset(Airports) and called it fl_delay

fl_delay = pd.concat([Delays, Airports], axis=1, ignore_index=False) 

In [58]:
fl_delay.head()

YEAR  MONTH  DayofMonth  DAY_OF_WEEK DEPARTURE_TIME  SCHEDULED_DEPARTURE  \
0  2008      1         3.0            4         2003.0                 1955   
1  2008      1         3.0            4          754.0                  735   
2  2008      1         3.0            4          628.0                  620   
3  2008      1         3.0            4         1829.0                 1755   
4  2008      1         3.0            4         1940.0                 1915   

  ARRIVAL_TIME  SCHEDULED_ARRIVAL                 AIRLINE  FLIGHT_NUMBER  ...  \
0       2211.0               2225  Southwest Airlines Co.            335  ...   
1       1002.0               1000  Southwest Airlines Co.           3231  ...   
2        804.0                750  Southwest Airlines Co.            448  ...   
3       1959.0               1925  Southwest Airlines Co.           3920  ...   
4       2121.0               2110  Southwest Airlines Co.            378  ...   

  AIRLINE_DELAY        id ident           type                  name  \
0           NaN    6523.0   00A       heliport     Total RF Heliport   
1           NaN  323361.0  00AA  small_airport  Aero B Ranch Airport   
2           NaN    6524.0  00AK  small_airport          Lowell Field   
3           NaN    6525.0  00AL  small_airport          Epps Airpark   
4           NaN  506791.0  00AN  small_airport  Katmai Lodge Airport   

  latitude_deg longitude_deg elevation_ft  iso_country iso_region  
0    40.070985    -74.933689         11.0           US      US-PA  
1    38.704022   -101.473911       3435.0           US      US-KS  
2    59.947733   -151.692524        450.0           US      US-AK  
3    34.864799    -86.770302        820.0           US      US-AL  
4    59.093287   -156.456699         80.0           US      US-AK  

[5 rows x 42 columns]

In [59]:
fl_delay.isnull().sum()

YEAR                         0
MONTH                        0
DayofMonth             5819079
DAY_OF_WEEK                  0
DEPARTURE_TIME               0
SCHEDULED_DEPARTURE          0
ARRIVAL_TIME                 0
SCHEDULED_ARRIVAL            0
AIRLINE                      0
FLIGHT_NUMBER                0
TAIL_NUMBER                  0
ELAPSED_TIME                 0
SCHEDULED_TIME               0
AIR_TIME                     0
ARRIVAL_DELAY                0
DEPARTURE_DELAY              0
ORIGIN_AIRPORT               0
DESTINATION_AIRPORT          0
DISTANCE                     0
TAXI_IN                      0
TAXI_OUT                     0
CANCELLED                    0
DIVERTED                     0
AIR_SYSTEM_DELAY             0
WEATHER_DELAY                0
NAS_DELAY              5819079
SECURITY_DELAY               0
LATE_AIRCRAFT_DELAY          0
DAY                    1936758
WHEELS_OFF             1936758
WHEELS_ON              1936758
CANCELLATION_REASON    1936758
AIRLINE_

In [60]:
fl_delay = fl_delay.fillna('0')

In [61]:
fl_delay.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7755837 entries, 0 to 7755836
Data columns (total 42 columns):
 #   Column               Dtype 
---  ------               ----- 
 0   YEAR                 int64 
 1   MONTH                int64 
 2   DayofMonth           object
 3   DAY_OF_WEEK          int64 
 4   DEPARTURE_TIME       object
 5   SCHEDULED_DEPARTURE  int64 
 6   ARRIVAL_TIME         object
 7   SCHEDULED_ARRIVAL    int64 
 8   AIRLINE              object
 9   FLIGHT_NUMBER        int64 
 10  TAIL_NUMBER          object
 11  ELAPSED_TIME         object
 12  SCHEDULED_TIME       object
 13  AIR_TIME             object
 14  ARRIVAL_DELAY        object
 15  DEPARTURE_DELAY      object
 16  ORIGIN_AIRPORT       object
 17  DESTINATION_AIRPORT  object
 18  DISTANCE             int64 
 19  TAXI_IN              object
 20  TAXI_OUT             object
 21  CANCELLED            int64 
 22  DIVERTED             int64 
 23  AIR_SYSTEM_DELAY     object
 24  WEATHER_DELAY        obj

This final merge changed most of the datatypes into object. Some object data types that were originally int, must first be converted into float first before they are converted back into the original int form.
These are the 'ARRIVAL_TIME'and 'DEPARTURE_TIME'columns.

In [62]:
fl_delay['ARRIVAL_TIME']= fl_delay['ARRIVAL_TIME'].astype('float')

In [63]:
fl_delay['DEPARTURE_TIME']= fl_delay['DEPARTURE_TIME'].astype('float')

In [64]:
fl_delay['ARRIVAL_TIME']= fl_delay['ARRIVAL_TIME'].astype('int')

In [65]:
fl_delay['DEPARTURE_TIME']= fl_delay['DEPARTURE_TIME'].astype('int')

The 'DAY' and 'id' columns can be converted straight away from object to int.

In [66]:
fl_delay['DAY']= fl_delay['DAY'].astype('int')

In [67]:
fl_delay['id']= fl_delay['id'].astype('int')

In [68]:
fl_delay.head()

YEAR  MONTH DayofMonth  DAY_OF_WEEK  DEPARTURE_TIME  SCHEDULED_DEPARTURE  \
0  2008      1        3.0            4            2003                 1955   
1  2008      1        3.0            4             754                  735   
2  2008      1        3.0            4             628                  620   
3  2008      1        3.0            4            1829                 1755   
4  2008      1        3.0            4            1940                 1915   

   ARRIVAL_TIME  SCHEDULED_ARRIVAL                 AIRLINE  FLIGHT_NUMBER  \
0          2211               2225  Southwest Airlines Co.            335   
1          1002               1000  Southwest Airlines Co.           3231   
2           804                750  Southwest Airlines Co.            448   
3          1959               1925  Southwest Airlines Co.           3920   
4          2121               2110  Southwest Airlines Co.            378   

   ... AIRLINE_DELAY      id ident           type                  name  \
0  ...             0    6523   00A       heliport     Total RF Heliport   
1  ...             0  323361  00AA  small_airport  Aero B Ranch Airport   
2  ...             0    6524  00AK  small_airport          Lowell Field   
3  ...             0    6525  00AL  small_airport          Epps Airpark   
4  ...             0  506791  00AN  small_airport  Katmai Lodge Airport   

  latitude_deg longitude_deg elevation_ft  iso_country iso_region  
0    40.070985    -74.933689         11.0           US      US-PA  
1    38.704022   -101.473911       3435.0           US      US-KS  
2    59.947733   -151.692524        450.0           US      US-AK  
3    34.864799    -86.770302        820.0           US      US-AL  
4    59.093287   -156.456699         80.0           US      US-AK  

[5 rows x 42 columns]

The following object data types must be converted back into the original float form.

In [69]:
# Converting columns back to float

fl_delay['DEPARTURE_DELAY']= fl_delay['DEPARTURE_DELAY'].astype('float32')

In [70]:
fl_delay['TAXI_OUT']= fl_delay['TAXI_OUT'].astype('float32')

In [71]:
fl_delay['WHEELS_OFF']= fl_delay['WHEELS_OFF'].astype('float32')

In [72]:
fl_delay['SCHEDULED_TIME'] =fl_delay['SCHEDULED_TIME'].astype('float32')

In [73]:
fl_delay['ELAPSED_TIME'] = fl_delay['ELAPSED_TIME'].astype('float32')

In [74]:
fl_delay['AIR_TIME']=fl_delay['AIR_TIME'].astype('float32')

In [75]:
fl_delay['WHEELS_ON']= fl_delay['WHEELS_ON'].astype('float32')

In [76]:
fl_delay['TAXI_IN']= fl_delay['TAXI_IN'].astype('float32')

In [77]:
fl_delay['ARRIVAL_TIME']=fl_delay['ARRIVAL_TIME'].astype('float32')

In [78]:
fl_delay['ARRIVAL_DELAY']= fl_delay['ARRIVAL_DELAY'].astype('float32')

In [79]:
fl_delay['AIR_SYSTEM_DELAY']=fl_delay['AIR_SYSTEM_DELAY'].astype('float32')

In [80]:
fl_delay['SECURITY_DELAY']=fl_delay['SECURITY_DELAY'].astype('float32')

In [81]:
fl_delay['AIRLINE_DELAY']=fl_delay['AIRLINE_DELAY'].astype('float32')

In [82]:
fl_delay['LATE_AIRCRAFT_DELAY']=fl_delay['LATE_AIRCRAFT_DELAY'].astype('float32')

In [83]:
fl_delay['WEATHER_DELAY']=fl_delay['WEATHER_DELAY'].astype('float32')

In [84]:
fl_delay['NAS_DELAY']=fl_delay['NAS_DELAY'].astype('float32')

In [85]:
fl_delay['latitude_deg']=fl_delay['latitude_deg'].astype('float32')

In [86]:
fl_delay['longitude_deg']= fl_delay['longitude_deg'].astype('float32')

In [87]:
fl_delay['elevation_ft'] =fl_delay['elevation_ft'].astype('float32')

In [88]:
# Checking if datatype changes have been effected.

fl_delay.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7755837 entries, 0 to 7755836
Data columns (total 42 columns):
 #   Column               Dtype  
---  ------               -----  
 0   YEAR                 int64  
 1   MONTH                int64  
 2   DayofMonth           object 
 3   DAY_OF_WEEK          int64  
 4   DEPARTURE_TIME       int32  
 5   SCHEDULED_DEPARTURE  int64  
 6   ARRIVAL_TIME         float32
 7   SCHEDULED_ARRIVAL    int64  
 8   AIRLINE              object 
 9   FLIGHT_NUMBER        int64  
 10  TAIL_NUMBER          object 
 11  ELAPSED_TIME         float32
 12  SCHEDULED_TIME       float32
 13  AIR_TIME             float32
 14  ARRIVAL_DELAY        float32
 15  DEPARTURE_DELAY      float32
 16  ORIGIN_AIRPORT       object 
 17  DESTINATION_AIRPORT  object 
 18  DISTANCE             int64  
 19  TAXI_IN              float32
 20  TAXI_OUT             float32
 21  CANCELLED            int64  
 22  DIVERTED             int64  
 23  AIR_SYSTEM_DELAY     float32
 24

In [89]:
# Checking final shape of cleaned dataset after the final merge with supplementary dataset

fl_delay.shape

(7755837, 42)

In [90]:
fl_delay.head()

YEAR  MONTH DayofMonth  DAY_OF_WEEK  DEPARTURE_TIME  SCHEDULED_DEPARTURE  \
0  2008      1        3.0            4            2003                 1955   
1  2008      1        3.0            4             754                  735   
2  2008      1        3.0            4             628                  620   
3  2008      1        3.0            4            1829                 1755   
4  2008      1        3.0            4            1940                 1915   

   ARRIVAL_TIME  SCHEDULED_ARRIVAL                 AIRLINE  FLIGHT_NUMBER  \
0        2211.0               2225  Southwest Airlines Co.            335   
1        1002.0               1000  Southwest Airlines Co.           3231   
2         804.0                750  Southwest Airlines Co.            448   
3        1959.0               1925  Southwest Airlines Co.           3920   
4        2121.0               2110  Southwest Airlines Co.            378   

   ... AIRLINE_DELAY      id  ident           type                  name  \
0  ...           0.0    6523    00A       heliport     Total RF Heliport   
1  ...           0.0  323361   00AA  small_airport  Aero B Ranch Airport   
2  ...           0.0    6524   00AK  small_airport          Lowell Field   
3  ...           0.0    6525   00AL  small_airport          Epps Airpark   
4  ...           0.0  506791   00AN  small_airport  Katmai Lodge Airport   

   latitude_deg longitude_deg elevation_ft  iso_country  iso_region  
0     40.070984    -74.933685         11.0           US       US-PA  
1     38.704021   -101.473907       3435.0           US       US-KS  
2     59.947735   -151.692520        450.0           US       US-AK  
3     34.864799    -86.770302        820.0           US       US-AL  
4     59.093288   -156.456696         80.0           US       US-AK  

[5 rows x 42 columns]

In [91]:
# Saved Cleaned dataset

fl_delay.to_csv('fl_delay.csv', sep=',')

**SUMMARY**


The datasets of flight delays in 2008 and 2015 were cleaned separately. 

Dataset of 2008 named,Delay_2008, started with 1936758 rows and 30 columns.
Missing columns were handled by dropping irrelevant columns like Cancellationcode column. Missing values were replaced with 0.


Dataset of 2015 named, Delay_2015 started with 5819079 rows and 31 columns.
Missing values were replaced with 0.

Columns with abbreviations like AIRLINE, ORIGIN_AIRPORT and DESTINATION_AIRPORT were repalced with full names from the airlines.csv and airports.csv datasets also downloaded from the Kaggle flight datasets page.


The supplementary dataset callled Airports started with 80557 rows and 18 columns.
Dropped columns were:
1. gps_code column beacause it has missing values but it also has the same information as the ident column which has no missing values
2. home_link, wikipedia_link because information about where data was scrapped is not necessary
3. local_code because it has missing values but it same information as ident column 
4. iata_code because is has a lot of of values missing which can not be replaced
5. iso_country because it has missing values but missing information is found in iso_region
6. municipality because it is irrelevant. iso_region has the required information about where the airport is located
7. scheduled_service because its is also no relevant to this project.

Missing values in iso_country column were replaced with country value from iso_region column.
missing values in elevation_ft column were replaced with median value of the column. Missing values in municipality column were replaced with 0. Data on only US Airports were selected and saved under the name Airports. Continent column was also dropped because it was no longer relevant.

Columns in Delay_2008 were renamed to match columns in Delay_2015. These two datasets were merged with Delay_2008 on top of Delay_2015. This new dataset was named 'Delays'. The cleaned supplementary dataset; 'Airports' was subsequently also merged with Delays using a left join. All missing columns were filled with 0 after the merge.

Columns with float data type were converted to int data type. 

Cleaned dataset was saved as fl_delay. It has 7755837 rows and 42 columns.